tools/ci.sh

```diff
@@ -2,41 +2,153 @@
-rm -rf fabric/api/fabric-client-kv-org*
-cd ./fabric/api/
-rm -rf ./node_modules/websocket/.git/
-npm install
-cd ../../
-npm run fabric:down
-npm run fabric
-cd ./quorum/api/
-rm -rf ./node_modules/websocket/.git/
-npm install
-cd ../../
-npm run quorum:down
-npm run quorum
-npm run fed:build
-npm run fed:quorum
-docker images
-while docker ps | grep "starting\|unhealthy"; do sleep 15; echo; date; done
-docker ps -a
-npm run fabric:down
-npm run quorum:down
-cd ../..
-#  - npm run test:bc - TODO: fix it, broken now
+startedAt=`date +%s`
+baseDir=`pwd`
+
+function mainTask()
+{
+  set -euxo pipefail
+
+  if ! [ -x "$(command -v lscpu)" ]; then
+    echo 'lscpu is not installed, skipping...'
+  else
+    lscpu
+  fi
+
+  if ! [ -x "$(command -v lsmem)" ]; then
+    echo 'lsmem is not installed, skipping...'
+  else
+    lsmem
+  fi
+
+  docker --version
+  docker-compose --version
+  node --version
+  npm --version
+  java -version || true
+
+  ### COMMON
+
+  rm -rf ./examples/simple-asset-transfer/fabric/api/node_modules
+  rm -rf ./examples/simple-asset-transfer/quorum/api/node_modules
+  rm -rf ./examples/simple-asset-transfer/node_modules
+  rm -rf ./node_modules
+
+  npm install
+  npm run test
+
+  cd examples/simple-asset-transfer
+  npm run fed:quorum:down
+  npm run fabric:down
+  npm run quorum:down
+  npm run quorum:api:down
+
+  npm install
+  rm -rf ./node_modules/websocket/.git/
+
+  ### FABRIC
+  cd ./fabric/api/
+  rm -rf ./fabric-client-kv-org*
+  rm -rf ./node_modules/websocket/.git/
+  npm install
+  cd ../../
+  npm run fabric
+
+  ### QUORUM
+
+  # Create docker network if not already in place
+  docker network inspect quorum_quorum-examples-net > /dev/null || docker network create --gateway=172.16.239.1 --subnet=172.16.239.0/24 quorum_quorum-examples-net
+  cd ./quorum/api/
+  rm -rf ./node_modules/websocket/.git/
+  npm install
+  cd ../../
+  npm run quorum
+  npm run quorum:api:build
+  npm run quorum:api
+  npm run fed:build
+  npm run fed:quorum
+  docker images
+
+  # If enough time have passed and there are still containers not ready then
+  # just assume that they are in a crash loop and abort CI run.
+  iterationCount=1
+  while docker ps | grep "starting\|unhealthy"; do
+    if [ "$iterationCount" -gt 20 ]; then
+      false;
+    fi
+    iterationCount=$[$iterationCount +1]
+    sleep 15; echo; date;
+  done
+
+  docker ps -a
+  sleep 120
+  npm run test:bc
+  npm run fed:quorum:down
+  npm run fabric:down
+  npm run quorum:down
+  npm run quorum:api:down
+  cd ../..
+
+  dumpAllLogs
+
+  endedAt=`date +%s`
+  runtime=$((endedAt-startedAt))
+  echo "$(date -Iseconds) [CI] SUCCESS - runtime=$runtime seconds."
+  exit 0
+}
+
+function onTaskFailure()
+{
+  set +e # do not crash process upon individual command failures
+
+  dumpAllLogs
+
+  endedAt=`date +%s`
+  runtime=$((endedAt-startedAt))
+  echo "$(date -Iseconds) [CI] FAILURE - runtime=$runtime seconds."
+  exit 1
+}
+
+function dumpAllLogs()
```

Comment:
you can use `after_script` step in .travis.yml for this. 
In fact, I believe this script should stay simple, easy to read and modify. Otherwise this script will soon need tests for itself. This dump logic could be moved to another script and be called from `after_script` step that runs if both script succeeded or failed.

Reply:
I wouldn't mind putting it in a different file, but using the travis-only features breaks down the idea that I can run the exact same CI locally that will run remotely on my branch once I open the PR. Tried looking for some tooling from Travis that allows me to "execute" the travis.yml the same way they do, but this is not something they opened up to the public (I guess it makes sense, a chunk of their value add would be given away that way).

Reply:
✅ 32864944bed098a1573df188376095f1067941e8